src/charm.py

```diff
@@ -509,73 +525,147 @@ def relabeling_config(self) -> list:
+                for key in snap_yaml["apps"].keys():
+                    snap_app_name = key  # Just use any app.
+                    break
+                log_dirs = self._evaluate_log_paths(
+                    paths=log_dirs, snap=endpoint.owner, app=snap_app_name
+                )
+                for path in log_dirs:
+                    job = self._snap_plug_job(
+                        endpoint.owner,
+                        f"{path}/**",
+                        topology.application,
+                        str(topology.unit),
+                        self._path_label(path),
+                    )
+                    shared_logs_configs.append(job)
 
-            job["relabel_configs"] = [
-                {
-                    "source_labels": ["__path__"],
-                    "target_label": "path",
-                    "replacement": fstab_entry.relative_target,
+        else:
+            endpoint_owners = {
+                endpoint.owner: {
+                    "juju_application": topology.application,
+                    "juju_unit": topology.unit,
                 }
-            ]
-
-            shared_logs_configs.append(job)
+                for endpoint, topology in self._cos.snap_log_endpoints_with_topology
+            }
+            for fstab_entry in agent_fstab.entries:
+                if fstab_entry.owner not in endpoint_owners.keys():
+                    continue
+
+                target_path = (
+                    f"{fstab_entry.target}/**"
+                    if fstab_entry
+                    else "/snap/grafana-agent/current/shared-logs/**"
+                )
+
+                job = self._snap_plug_job(
+                    fstab_entry.owner,
+                    target_path,
+                    endpoint_owners[fstab_entry.owner]["juju_application"],
+                    endpoint_owners[fstab_entry.owner]["juju_unit"],
+                    fstab_entry.relative_target,
+                )
+                shared_logs_configs.append(job)
 
         return shared_logs_configs
 
     def _connect_logging_snap_endpoints(self):
-        for plug in self._cos.snap_log_endpoints:
-            try:
-                self.snap.connect("logs", service=plug.owner, slot=plug.name)
-            except snap.SnapError as e:
-                logger.error(f"error connecting plug {plug} to grafana-agent:logs")
-                logger.error(e.message)
+        # We need to run _verify_snap_track so we make sure we have refreshed BEFORE connecting.
+        self._verify_snap_track()
+        if not self.config["classic_snap"]:
+            for plug in self._cos.snap_log_endpoints:
+                try:
+                    self.snap.connect("logs", service=plug.owner, slot=plug.name)
+                except snap.SnapError as e:
+                    logger.error(f"error connecting plug {plug} to grafana-agent:logs")
+                    logger.error(e.message)
```

Comment:
Should we add to the log line: "Re-relate to the charm to retry"? (if it's correct)